0000-qiskit-pulse-schedule-relative-timing.md

```diff
@@ -0,0 +1,105 @@
+
+ - Make appending two schedules a much simpler operation as the dependency of instructions will be specified by their order on operands
+ - This will allow a simple transformation to a DAG representation of a basic pulse schedule block
+ - Have positive performance impacts.
+ - Have implications for the current pulse API.
+
+## User Benefit
+ - Pulse users (primarily, experimental physicists interested in using IBM backends) should get a perforance improvement.
+ - When we begin implementing control flow around Schedules, we will find this new implementation much easier to work with. This benefits frontend and backend developers, which in turn will give our end users new features sooner.
+
+## Design Proposal
+
+<!-- This is the focus of the document. Explain the proposal from the perspective of
+educating another user on the proposed features.
+
+This generally means:
+- Introducing new concepts and nomenclature
+- Using examples to introduce new features
+- Implementation and Migration path with associated concerns
+- Communication of features and changes to users
+
+Focus on giving an overview of impact of the proposed changes to the target
+audience.
+
+Factors to consider:
+- Performance
+- Dependencies
+- Maintenance
+- Compatibility -->
+
+### Notation:
+I am going to use this notation here, where `self` is the `Schedule` initiating the operation in question:
+
+ - |C| : the total number of channels in `self`
+ - |N<sub>self</sub>| : the total number of instructions in `self`
+ - |N<sub>new</sub>| : the total number of instructions in another `Schedule` being added in some way to `self`
+ - max(|N<sub>C</sub>|) : the maximum number of instructions on any single channel within `self`
+
+### New Schedule methods
+
+| Method        | Runtime        |  Description     |
+|---------------|----------------|------------------|
+|`barrier(self)`| O(\|C\|) | Add `Delay`s across `channels` so that all of the `channels` are occupied until `self.duration(channels)`. |
+|`left_align(self)` | O(\|C\|) | This is the default behavior. We could have it return `self`, or we could remove any pre-pended `Delay`s. The latter is probably better, because `schedule.right_align().left_align()` should yield the original `schedule`. Should we also append `Delay`s? Likely. *Needs more thought* |
+|`center_align(self)` | O(\|C\|) | Pre- and ap-pend equally sized `Delay`s (how to split odd durations?) on each channel so that all channels are occupied until `self.duration()` |
+|`right_align(self)` | O(\|C\|) | Prepend `Delay`s on each channel so that all channels are occupied until `self.duration()`. Should we also remove trailing `Delay`s? |
+
+
+### API Changes
+
+\# TODO
+
+| Method        | Runtime        |  Change          |
+|---------------|----------------|------------------|
+| `append(self, schedule)` | O(\|N<sub>new</sub>\|) |  |
+|`insert(self, time, schedule)`| O(\|N<sub>new</sub>\| * \|N<sub>C</sub>\|) |  |
```

Comment:
How do you expect inserting in the middle of a schedule work? Verify that delays are present for all insertion locations?

Reply:
yes. it's not great, since delays are supposed to be blocking, but I don't think there's any other way to do this